### PR TITLE
Refactor AppProvider workflow and protocol

### LINKED ISSUE #991: UIURL not parsed in parseConfig of appprovider
found the issue while doing some simple unit tests, see https://github.com/cs3org/reva/pull/992

### PR DESCRIPTION
**Pull request by glpatcern**
This PR uses https://github.com/cs3org/cs3apis/pull/85 to implement a simpler protocol from the client to the Reva gateway, and a more enriched one from the gateway to the appprovider. For now, WOPI is still the only "driver" implementing the appprovider logic.

Also, `storageID` and `UIURL` are removed from the app-provider's config: `StorageID` is auto-resolved in the gateway service, whereas the `UIURL`'s doc was misleading: we need to decide whether to drop it entirely (and surely it won't stay in the config).

En passant, this fixes #991 and includes a few other changes and fixes concerning error handling, logging and documentation.

### CODE REVIEW

**Comment by labkode** [comment]
@glpatcern cs3apis merged, remember changelog

**Comment by labkode** [comment]
@glpatcern looks already pretty good

File: internal/grpc/interceptors/recovery/recovery.go [hunk 1]
```diff
@@ -45,8 +45,8 @@ func NewStream() grpc.StreamServerInterceptor {
 }
 
 func recoveryFunc(ctx context.Context, p interface{}) (err error) {
-	stack := debug.Stack()
+	debug.PrintStack()
```
Comment (ishank011) [inline internal/grpc/interceptors/recovery/recovery.go:48]: Nice!

File: internal/grpc/services/appprovider/appprovider.go [hunk 2]
```diff
@@ -154,50 +146,41 @@ func (s *service) OpenFileInAppProvider(ctx context.Context, req *providerpb.Ope
 		return nil, err
 	}
 
-	httpReq, err := rhttp.NewRequest(ctx, "GET", wopiurl+"wopi/iop/open", nil)
+	httpReq, err := rhttp.NewRequest(ctx, "GET", s.conf.WopiURL+"wopi/iop/open", nil)
```
Comment (ishank011) [inline internal/grpc/services/appprovider/appprovider.go:149]: Can we take care of missing trailing slashes here? Something like [this](https://github.com/cs3org/reva/blob/master/pkg/ocm/invite/manager/json/json.go#L210)

File: internal/grpc/interceptors/recovery/recovery.go (hunk 1 above)
Comment (glpatcern) [reply internal/grpc/interceptors/recovery/recovery.go:48]: ;-) this was needed as I got a full stack printed with literal `\n`'s all over!

File: internal/grpc/services/appprovider/appprovider.go (hunk 2 above)
Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go:149]: Sure, makes sense

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go]: How does it look like? I note that the block with the HTTP request to `/wopi/cbox/endpoints` should be moved elsewhere, in some kind of init of the service. Maybe I could move it to the `init()` function? or where else?

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go]: Answering to myself, in fact I found a way to do that with a separate function. This appprovider will eventually be the WOPI driver, so it's OK to have it like this - commit to follow.

File: internal/grpc/services/appprovider/appprovider.go [hunk 3]
```diff
@@ -48,8 +49,9 @@ func init() {
 }
 
 type service struct {
-	provider app.Provider
-	conf     *config
+	provider   app.Provider
+	conf       *config
+	appsURLMap map[string]interface{}
```
Comment (ishank011) [inline internal/grpc/services/appprovider/appprovider.go:54]: This needs to be protected by a lock. Multiple threads can try to fill this at the same time, and even though they would probably be writing the same data, there's no ordering in a map, so we might end up with garbage values.

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go:54]: Well spotted, too much Python coding ;-) I'll fix that.

Comment (ishank011) [reply internal/grpc/services/appprovider/appprovider.go:54]: You can consider using a sync.Map like done [here](https://github.com/cs3org/reva/blob/master/pkg/ocm/provider/authorizer/json/json.go#L83).

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go:55]: In fact, given the usage (write once, read many), for now I went for a standard `mutex.Lock()` approach. Just committed.

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go:55]: (true that once a regular refresh is implemented, the best option is a sync.Map)

Comment (ishank011) [reply internal/grpc/services/appprovider/appprovider.go:55]: This won't work. The lock is local to each goroutine, so it won't know what it actually has to take a lock on. Ref: https://play.golang.org/p/RZfwVALEvRn. You'd have to add this lock to the service struct itself.

And why not use sync.Map in the first place since we agree it would be the better option in the long run?

Comment (glpatcern) [reply internal/grpc/services/appprovider/appprovider.go]: (Thanks for your patience, I genuinely assumed Go had some higher level constructs, like mutexes being stored in some common memory. Instead, it's pretty much like C/C++, good to know)

OK, I gave it a try with sync.Map but it's not a good fit as it's not a real map; and for the long run, we want to invalidate its content and call that function again to fill it. So at that point I'll implement a proper multi-read/single-write lock for it, but for this PR, I came back to all local variables. I guess it's better to move on for now and leave that as a separate task.

File: internal/grpc/interceptors/recovery/recovery.go (hunk 1 above)
Comment (labkode) [inline internal/grpc/interceptors/recovery/recovery.go:48]: @glpatcern can you revert this change? Even though is useful it will break log parsing. Let's create a github issue to proper address it.

Comment (glpatcern) [reply internal/grpc/interceptors/recovery/recovery.go:48]: Interestingly, Ishank marked this "nice", and was motivated by my comment above: "I got a full stack printed with literal \n's all over!". Why would this break log parsing? Because it's multiline?

Comment (glpatcern) [reply internal/grpc/interceptors/recovery/recovery.go:48]: Opened #1054 to track this.
